### DISCUSSION TITLE
[Question] Reuse Existing SimulationApp

### DISCUSSION THREAD

**Discussion by davis-openai** [isaac-sim/IsaacLab#615] [Q&A] [answered]
### Question

Is it possible to reusue a SimulationApp that was previously started instead of launching a new one with AppLauncher each time I run a new workflow? Perhaps I'm missing something here, but launching Isaac Sim every time takes a significant amount of time. Thank you!

**Comment by Mayankm96** [comment, marked as answer]
There are [two main workflows](https://docs.omniverse.nvidia.com/isaacsim/latest/introductory_tutorials/tutorial_intro_workflows.html) for using Omniverse:

- **Extension**: where you launch the simulator once and can add callbacks to it through Python scripting. This allows hot reloading of the code without needing to close the simulation app itself.
- **Standalone**: where you launch a Python script that launches the simulator and provides you complete control over the simulation app.

In most RL libraries, the agent-environment running loop is designed for the "standalone" workflow where the runner has complete control over the environment stepping. Making this work for an "extension" workflow is not impossible but will require writing some queueing mechanisms to send and receive states from the simulator.

We tried this out with OmniIsaacGymEnvs, and there is a multi-threaded implementation for [this purpose](https://github.com/isaac-sim/OmniIsaacGymEnvs?tab=readme-ov-file#extension-workflow). However, the added code complexity was not so appealing and we didn't end up porting it over yet. Instead, we have shifted the focus more on reducing the startup time to make the simulation app as lightweight as possible. This may take some time, but we hope to reduce this number significantly this year.

**Reply by davis-openai** [reply]
So just to clarify, there is currently no way with Isaac Lab to start a simulation context and then load a different environment later?

**Reply by bikcrum** [reply]
@Mayankm96 I’ve run into an issue where, once AppLauncher is initialized, code changes (e.g., in env_cfg) are not reflected even after reloading the module. Interestingly, the class content does appear updated, but the behavior doesn’t change, suggesting that AppLauncher might be freezing the Python module and retaining that initial state throughout its lifecycle.

I’m able to reload and run other tasks (registered through gym) without any issue. It's specifically the code updates that seem to be ignored.

I’ve shared more details on what I’m trying to achieve in this [discussion thread](https://github.com/isaac-sim/IsaacLab/discussions/2349). I’d appreciate any suggestions or workarounds you might have.
